### ISSUE TITLE
LXD servers in cluster, TCP sessions not clossing

### ISSUE THREAD

**Issue by ghost** [lxc/lxd#5179]
* Distribution: ubuntu 
 * Distribution version:18.04 LTS
 * The output of "lxc info:
api_extensions:
- storage_zfs_remove_snapshots
- container_host_shutdown_timeout
- container_stop_priority
- container_syscall_filtering
- auth_pki
- container_last_used_at
- etag
- patch
- usb_devices
- https_allowed_credentials
- image_compression_algorithm
- directory_manipulation
- container_cpu_time
- storage_zfs_use_refquota
- storage_lvm_mount_options
- network
- profile_usedby
- container_push
- container_exec_recording
- certificate_update
- container_exec_signal_handling
- gpu_devices
- container_image_properties
- migration_progress
- id_map
- network_firewall_filtering
- network_routes
- storage
- file_delete
- file_append
- network_dhcp_expiry
- storage_lvm_vg_rename
- storage_lvm_thinpool_rename
- network_vlan
- image_create_aliases
- container_stateless_copy
- container_only_migration
- storage_zfs_clone_copy
- unix_device_rename
- storage_lvm_use_thinpool
- storage_rsync_bwlimit
- network_vxlan_interface
- storage_btrfs_mount_options
- entity_description
- image_force_refresh
- storage_lvm_lv_resizing
- id_map_base
- file_symlinks
- container_push_target
- network_vlan_physical
- storage_images_delete
- container_edit_metadata
- container_snapshot_stateful_migration
- storage_driver_ceph
- storage_ceph_user_name
- resource_limits
- storage_volatile_initial_source
- storage_ceph_force_osd_reuse
- storage_block_filesystem_btrfs
- resources
- kernel_limits
- storage_api_volume_rename
- macaroon_authentication
- network_sriov
- console
- restrict_devlxd
- migration_pre_copy
- infiniband
- maas_network
- devlxd_events
- proxy
- network_dhcp_gateway
- file_get_symlink
- network_leases
- unix_device_hotplug
- storage_api_local_volume_handling
- operation_description
- clustering
- event_lifecycle
- storage_api_remote_volume_handling
- nvidia_runtime
api_status: stable
api_version: "1.0"
auth: trusted
public: false
auth_methods:

  architectures:
  - x86_64
  - i686

 driver: lxc
  driver_version: 3.0.1
  kernel: Linux
  kernel_architecture: x86_64
  kernel_version: 4.15.0-36-generic
  server: lxd
  server_pid: 16674
  server_version: 3.0.1
  storage: dir
  storage_version: "1"

# Issue description

Servers are having huge numbers of established TCP sessions. Over week reaching 1000 established connections. Reloading socket with:
systemctl stop lxd.socket
systemctl start lxd.socket
Helps do lower number of connectios.

# Steps to reproduce
 1. use LXD
 2. every night do backups with lxc snapshot, lxc publish, lxc export
 3. End of week TCP established sessions reach up to 1000 and more.

# Information to attach

netstat -antpl | grep 8443 | wc -l   2 hours after socket reload
148 sessions

![capture](https://user-images.githubusercontent.com/39885322/47031828-6fb23c80-d179-11e8-8747-144d04ec242f.PNG)
We testing LXD on 2 servers each has up to 3 containers

**Comment by vudududu** [comment]
Servers are mostly on default config form install

**Comment by stgraber** [comment]
@freeekanayaka can you take a look?

On my test cluster, I'm seeing over 250 ESTABLISHED connections despite not running any queries against it at the moment.

Unlike @vudududu my cluster is running on up to date 3.6 so it's not an issue that's gotten fixed since.

**Comment by vudududu** [comment]
lxc list command after socket reload generated:
netstat -antpl |grep 8443 |grep ESTABLISHED |grep lxd | wc -l
46

**Comment by freeekanayaka** [comment]
@vudududu: are you using the clustering feature?

**Comment by vudududu** [comment]
@freeekanayaka if you are asking if servers are in cluster than yes. I set it up during lxc init

**Comment by freeekanayaka** [comment]
@vudududu just to avoid confusion, if servers are in a cluster, why do you need to use ```lxc publish``` in your daily backup? Are you pushing snapshots to some lxd instance external to the cluster?

**Comment by vudududu** [comment]
@freeekanayaka  Container is running. I make a snapshot. Publish snapshot to a image and export image to netapp for storage.
We have around 35 servers and about 150 lxc containers. Now we must move all of this to lxd but we must keep backups of containers as requirement.
Using google I found out its the only way to get full system backup that can be moved and restored in development servers

**Comment by freeekanayaka** [comment]
Let's keep an eye on this one, because I'm really not sure that the above change fixes the issue.

@vudududu if you could try to switch to the candidate snap channel and check if this works for you, that'd be great.

**Comment by stgraber** [comment]
Re-opening as I'm still seeing a much higher than expected number of connections on my test cluster.

I'm running a 4 nodes cluster with 1 client interacting with it and have a lot of ESTABLISHED connections still:
```
root@infra02:~# netstat -taunp | grep lxd | grep ESTAB | wc -l
369
```

**Comment by nanjj** [comment]
@stgraber 
From https://golang.org/pkg/net/http/#Client:
```
The Client's Transport typically has internal state (cached TCP connections), 
so Clients should be reused instead of created as needed. 
```
while, at https://github.com/lxc/lxd/blob/master/client/util.go#L91:
```
if client == nil {
		client = &http.Client{}
}
```
we create new http client. 

I suggest to reuse via global var:
```
var DefaultClient = &Client{}
...
if client == nil {
		client = DefaultClient
}
```

**Comment by freeekanayaka** [comment]
Interesting take, probably worth trying.

**Comment by stgraber** [comment]
A global var wouldn't work though as the client actually differs on a per-remote basis as we build and set a custom transport for each of those.

If that's in fact the source of the issue for our internal use in a cluster, it should be fixable by ensuring that all cluster connections use the args.HTTPClient option of the client to pass an existing http client for the particular remote server they're interacting with (as using a global one would lead to its transport being rewritten constantly leading to failures).

This also doesn't really explain what we're seeing here as I would still expect clients to get freed, releasing the connections when they go out of scope rather than keeping all those connections active.

**Comment by nanjj** [comment]
@stgraber Look the discussion here: 
https://github.com/golang/go/issues/24739
Which can explain the leak.

**Comment by stgraber** [comment]
@vudududu are you still seeing this problem regularly? I've not noticed it with my clusters for a while now.

**Comment by vudududu** [comment]
Hi, all. Sorry cant test new versions since we run with apt install lxd, management push movement to production , currently cant move to snap.  Made cron task to clean tcp sessions from socket.
HOSTNAME=$(hostname)
systemctl stop lxd.socket
systemctl start lxd.socket
sleep 10s
lxc list

For enterprise sessions and clustering in LXD is important and without  them  LXD cant move fully to enterprise . After migration  I will try to test snap version ASAP.
Thank you for help.

**Comment by aaronvegh** [comment]
aaron@codewerks-alpha:~$ lxd --version
3.15

aaron@codewerks-alpha:~$ sudo netstat -antpl |grep 8443 |grep ESTABLISHED |grep lxd | wc -l
641

As of now my LXD installation is non-responsive. I only have one running container on my 3-node cluster but I can't work with it at all. I'm going to try rebooting now and hope it clears this out, but I'm certainly hoping for a solution to this!

**Comment by vudududu** [comment]
@aaronvegh  you can make script and  cronjob
#!/bin/bash
HOSTNAME=$(hostname)
systemctl stop lxd.socket
systemctl start lxd.socket
sleep 10s
lxc list

Starting first node, ending with last. 
We do this every morning we have 2 cluster each with 24 nodes. Each node holds around 30~40 containers. We use first  3.0.0 LTS  version. It takes around 5~10 minutes for a full flush. But containers are running and services do not stop.

**Comment by aaronvegh** [comment]
I'd tried your commands last night but: `Failed to stop lxd.socket: Unit lxd.socket not loaded.`

I settled for stopping all LXD services I could find in `ps aux`, but that wasn't effective.

Currently:

aaron@codewerks-alpha:~$ sudo netstat -antpl |grep 8443 |grep ESTABLISHED |grep lxd | wc -l
89782

**Comment by vudududu** [comment]
@aaronvegh forgot to mention we do not use snap. so that script for non snap lxd. on snap it is a bit different.

**Comment by vudududu** [comment]
@aaronvegh I would suggest looking for information on https://discuss.linuxcontainers.org. If I remember correctly, I googled  around and found out  what must be reloaded on non snap lxd and on snap lxd

**Comment by aaronvegh** [comment]
I used the command `sudo systemctl kill snap.lxd.daemon` and that improves the number of connections, but it continues to massively climb and `lxc` remains unresponsive:

aaron@codewerks-alpha:~$ sudo netstat -antpl |grep 8443 |grep ESTABLISHED |grep lxd | wc -l
29
aaron@codewerks-alpha:~$ lxc list

**Comment by vudududu** [comment]
@aaronvegh Try looking for snap.lxd.socket.  on 3.0.X snap it was like this if I remmeber correctly. Then systemctl stop snap.lxd.socket and check connections if they dropped do systemctl start snap.lxd.socket

**Comment by aaronvegh** [comment]
@vudududu I appreciate your trying to help, btw. :-)

I found a few relevant services in the /etc/systemd/service directory:

```
aaron@codewerks-alpha:~$ ls /etc/systemd/system/snap*
/etc/systemd/system/snap-core-7169.mount  
/etc/systemd/system/snap-lxd-11270.mount       
/etc/systemd/system/snap.lxd.daemon.unix.socket
/etc/systemd/system/snap-core-7270.mount 
 /etc/systemd/system/snap.lxd.activate.service
/etc/systemd/system/snap-lxd-11254.mount  
/etc/systemd/system/snap.lxd.daemon.service
```

I tried killing and restarting `snap.lxd.daemon`, `snap.lxd.daemon.unix.socket` and `snap.lxd.activate`. This does blow away the connections, but it immediately starts exploding again, and even with a "low" number immediately after restarting, I cannot get `lxc list` to return anything.

**Comment by rosbeef** [comment]
hello,
I'm in the same case. I just explain my situation .
i'm with snap package 3.15 under ubuntu 18.04 armhf trying to run a self hosted multisite cluster.
LXD cluster is running over a ipv6 tap interface mounted by [freelan](http://freelan.org/) so over udp/https. I have in minutes more than hundred tcp connections. 
I notes that sometimes udp packets takes more than the expected time to be "recieved" by the remote tunnel agent. 

My interpretation is that or lxd have no keep-alive tcp connection or lxd expect a response too quickly. 
If you need some logs tell me which one and i hope you can resolve this situation. 
Good luck

**Comment by freeekanayaka** [comment]
@rosbeef hm so are you running the the members of your LXD cluster in physically distant networks? That might be a problem, or at least something we didn't expect users to do. Essentially things might get brittle due to the higher latency and unreliability of the network.

**Comment by aaronvegh** [comment]
Not sure what the better venue for this discussion is, but I opened a discussion about this on the Linux Containers forum:

https://discuss.linuxcontainers.org/t/lxd-3-15-cluster-hangs-with-massive-connection-count/5341/5

**Comment by rosbeef** [comment]
@freeekanayaka my reply at @aaronvegh [lxc forum link](https://discuss.linuxcontainers.org/t/lxd-3-15-cluster-hangs-with-massive-connection-count/5341/5)

**Comment by rosbeef** [comment]
some logs from /var/snap/lxd/common/lxd/logs/lxd.log

i dont really understand this error : "Failed connecting to global database (attempt 8): failed to create dqlite connection: no available dqlite leader server found" 

> t=2019-07-27T13:10:28+0000 lvl=info msg="LXD 3.15 is starting in normal mode" path=/var/snap/lxd/common/lxd
> t=2019-07-27T13:10:28+0000 lvl=info msg="Kernel uid/gid map:" 
> t=2019-07-27T13:10:28+0000 lvl=info msg=" - u 0 0 4294967295" 
> t=2019-07-27T13:10:28+0000 lvl=info msg=" - g 0 0 4294967295" 
> t=2019-07-27T13:10:28+0000 lvl=info msg="Configured LXD uid/gid map:" 
> t=2019-07-27T13:10:28+0000 lvl=info msg=" - u 0 1000000 1000000000" 
> t=2019-07-27T13:10:28+0000 lvl=info msg=" - g 0 1000000 1000000000" 
> t=2019-07-27T13:10:28+0000 lvl=warn msg="AppArmor support has been disabled because of lack of kernel support" 
> t=2019-07-27T13:10:28+0000 lvl=warn msg="Couldn't find the CGroup pids controller, process limits will be ignored." 
> t=2019-07-27T13:10:28+0000 lvl=info msg="Kernel features:" 
> t=2019-07-27T13:10:28+0000 lvl=info msg=" - netnsid-based network retrieval: no" 
> t=2019-07-27T13:10:28+0000 lvl=info msg=" - uevent injection: no" 
> t=2019-07-27T13:10:28+0000 lvl=info msg=" - seccomp listener: no" 
> t=2019-07-27T13:10:28+0000 lvl=info msg=" - unprivileged file capabilities: yes" 
> t=2019-07-27T13:10:28+0000 lvl=info msg=" - shiftfs support: no" 
> t=2019-07-27T13:10:28+0000 lvl=info msg="Initializing local database" 
> t=2019-07-27T13:10:28+0000 lvl=info msg="Starting /dev/lxd handler:" 
> t=2019-07-27T13:10:28+0000 lvl=info msg=" - binding devlxd socket" socket=/var/snap/lxd/common/lxd/devlxd/sock
> t=2019-07-27T13:10:28+0000 lvl=info msg="REST API daemon:" 
> t=2019-07-27T13:10:28+0000 lvl=info msg=" - binding Unix socket" inherited=true socket=/var/snap/lxd/common/lxd/unix.socket
> t=2019-07-27T13:10:28+0000 lvl=info msg=" - binding TCP socket" socket=[2aa1::100:1]:8443
> t=2019-07-27T13:10:28+0000 lvl=info msg="Initializing global database" 
> t=2019-07-27T13:11:51+0000 lvl=warn msg="Failed connecting to global database (attempt 6): failed to create dqlite connection: no available dqlite leader server found" 
> t=2019-07-27T13:12:03+0000 lvl=warn msg="Failed connecting to global database (attempt 7): failed to create dqlite connection: no available dqlite leader server found" 
> t=2019-07-27T13:12:15+0000 lvl=warn msg="Failed connecting to global database (attempt 8): failed to create dqlite connection: no available dqlite leader server found" 
> t=2019-07-27T13:12:28+0000 lvl=warn msg="Failed connecting to global database (attempt 9): failed to create dqlite connection: no available dqlite leader server found" 
... 
> t=2019-07-27T13:34:00+0000 lvl=warn msg="Failed connecting to global database (attempt 115): failed to create dqlite connection: no available dqlite leader server found" 
> t=2019-07-27T13:35:02+0000 lvl=warn msg="Failed connecting to global database (attempt 120): failed to create dqlite connection: no available dqlite leader server found"

**Comment by stgraber** [comment]
Meant to close this a few weeks back when we fixed a few issues around our event handling code in the Go client. This was effectively causing event listeners to stick around when they shouldn't, accumulating fds in a way consistent with this report.

So closing this on the assumption that the problem will finally be gone in LXD 3.19.
If that's not the case, we'll re-open and that will likely indicate that we can't rely on the auto-disconnect logic and will need to manually place .Disconnect() calls (also a newly introduced function in 3.19) to disconnect wherever we may be holding onto event handlers.

**Comment by freeekanayaka** [comment]
A TCP connection leak was detected and fixed in PR #6750, not sure it's the same issue, but it's worth noting.
